### ISSUE TITLE
"autosubmit log recovery" processes

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1314]
In GitLab by @abaer-kit on May 31, 2024, 15:22

Hello @dbeltrankyl and @kinow,

### Autosubmit Version

4.1.x

### Which task has issues?

autosubmit run

### Summary

If one starts an experiment, additional processes are started with the names "autosubmit log recovery" and a platform name (one for each platform).
These seem to be related to the main process, but are not canceled if the process stops (e.g. due to a successful finishing of the run).

Is this intended?\
Are they stopped at some point?

Interestingly, these jobs can stop a gitlab-runner job that is running Autosubmit from terminating even after a successful run.

### Steps to reproduce

Run an experiment

### Relevant logs and/or screenshots(if applicable)

End of the log output (In the end I tried to stop them by interrupting them. However, this did not stop them):
```
2024-05-31 14:57:44,349 Backing up jobs_data...
2024-05-31 14:57:44,355 Jobs_data database backup completed.
2024-05-31 14:57:54,363 No more jobs to run.
2024-05-31 14:57:54,363 Waiting for all logs to be updated
2024-05-31 14:57:54,367 Backing up jobs_data...
2024-05-31 14:57:54,371 Jobs_data database backup completed.
2024-05-31 14:57:54,372 Run successful
2024-05-31 15:03:47,287 Autosubmit will interrupt at the next safe occasion
2024-05-31 15:03:55,883 Autosubmit will interrupt at the next safe occasion
```

**Comment by kinow** [comment]
Thanks for reporting it, @abaer-kit !

@dbeltrankyl , this might be related to #1304, and maybe the zombie processes are happening not only on LUMI?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 31, 2024, 15:38

Hello @abaer-kit , 

We will do additional fixes in 4.1.7 ( already on master https://earth.bsc.es/gitlab/es/autosubmit/-/merge_requests/436 )

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 31, 2024, 15:39

Ah wait, are you using master? because of the `autosubmit log recovery` name, they still don't exit?

**Comment by kinow** [comment]
In GitLab by @abaer-kit on May 31, 2024, 16:07

Yes, I am using the master, thus including the code from !436.

The `autosubmit run` process actually does exit. However, further jobs are still active:
![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/1031b0491fa3bf6083906fa3334c6cc4/image.png)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 31, 2024, 17:42

Hello @abaer-kit , 

Since I can't reproduce the issue on my laptop, could you help me to test it? 

Could you test the branch 4.1.7-zombies_alternative? I've added a function called terminate_child_process that should avoid that. It relies on getting the PID with a terminal command, for UNIX systems, it should work but I didn't test on Mac

It is called autosubmit run and autosubmit stop commands. 

The autosubmit run should be called with control + c, crashes, successful runs, and whenever a platform has an issue. 

Also it is retroactive, so people that have some zombie processor should also kill them if they rerun the experiment or use the stop command

**Comment by kinow** [comment]
In GitLab by @abaer-kit on May 31, 2024, 18:13

> Since I can't reproduce the issue on my laptop, could you help me to test it?

Sure!

For a first test run (with a failing job), the processes are indeed terminated. Interestingly, the one for "levante" at the end, but the one for "levante_login" immediately after startup (or maybe it is not started at all).

Further, the logs are not retrieved, I don't know, what the problem here is. The Autosubmit log is attached: [full_log.txt](https://earth.bsc.es/gitlab/es/autosubmit/uploads/747e0a259abe489818bc7b6fbd07d21f/full_log.txt)

I can probably do further testing from mid of next week on.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 3, 2024, 10:12

Hello @abaer-kit 

Thanks for testing it, 

I think that what happened is that the grep matched *levante* instead of levante. Now, it should match the platform name and nothing else, as I changed the grep to match " platform_name.lower() ", instead of "platform_name.lower()".

> Further, the logs are not retrieved,

Which ones, only the levante_login ones? or both. I'll test. If it is only levante_login I guess it is because the log recovery for the platform was killed due the grep.

Edit: 

```
-rw-rw-r-- 1 dbeltran dbeltran   22 Jun  3 10:21 a053_TESTZOMBIES_MN5.20240603102124.out
-rw-rw-r-- 1 dbeltran dbeltran  949 Jun  3 10:21 a053_TESTZOMBIES_MN5.20240603102124.err
drwxr-xr-x 2 dbeltran dbeltran  12K Jun  3 10:21 ./
-rw-rw-r-- 1 dbeltran dbeltran   22 Jun  3 10:21 a053_TESTZOMBIES_LOGIN.20240603102126.out
-rw-rw-r-- 1 dbeltran dbeltran  955 Jun  3 10:21 a053_TESTZOMBIES_LOGIN.20240603102126.err
```

In my side, using the last update, it is working, used the platforms: marenostrum5_login(ps) and marenostrum5

**Comment by kinow** [comment]
In GitLab by @abaer-kit on Jun 3, 2024, 10:36

> Which ones, only the levante_login ones? or both.

No logs are retrieved and the logs directory (`LOG_<expid>` in `tmp`) is not created. I will check with the latest version on Wednesday (Levante is on maintenance) to see if it works correctly then.

**Comment by kinow** [comment]
In GitLab by @abaer-kit on Jun 10, 2024, 16:30

Hello @dbeltrankyl 

finally, I managed to test this. With the latest version of `4.1.7-zombies_alternative` it works as expected. The processes start at the beginning and terminate at the end and all the logs are retrieved accordingly.

Thanks for this!

Note (I tested this by accident also :D ): with the option DISABLE_RECOVERY_THREADS set to True, the threads are not used (as expected).
